src/compiler/tsbuildPublic.ts

```diff
@@ -962,35 +983,31 @@ namespace ts {
             // Actual Emit
             const { host, compilerHost } = state;
             let resultFlags = BuildResultFlags.DeclarationOutputUnchanged;
-            let newestDeclarationFileContentChangedTime = minimumDate;
-            let anyDtsChanged = false;
+            const existingBuildInfo = state.buildInfoCache.get(projectPath)?.buildInfo || undefined;
             const emitterDiagnostics = createDiagnosticCollection();
             const emittedOutputs = new Map<Path, string>();
-            outputFiles.forEach(({ name, text, writeByteOrderMark }) => {
-                let priorChangeTime: Date | undefined;
-                if (!anyDtsChanged && isDeclarationFileName(name)) {
-                    // Check for unchanged .d.ts files
-                    if (host.fileExists(name) && state.readFileWithCache(name) === text) {
-                        priorChangeTime = host.getModifiedTime(name);
-                    }
-                    else {
+            const options = program.getCompilerOptions();
+            const isIncremental = isIncrementalCompilation(options);
+            let outputTimeStampMap: ESMap<Path, Date> | undefined;
+            let now: Date | undefined;
+            outputFiles.forEach(({ name, text, writeByteOrderMark, buildInfo }) => {
+                const path = toPath(state, name);
+                emittedOutputs.set(toPath(state, name), name);
+                if (buildInfo) {
+                    setBuildInfo(state, buildInfo, projectPath, options);
+                    if (buildInfo.program?.dtsChangeTime !== existingBuildInfo?.program?.dtsChangeTime) {
                         resultFlags &= ~BuildResultFlags.DeclarationOutputUnchanged;
-                        anyDtsChanged = true;
                     }
                 }
-
-                emittedOutputs.set(toPath(state, name), name);
                 writeFile(writeFileCallback ? { writeFile: writeFileCallback } : compilerHost, emitterDiagnostics, name, text, writeByteOrderMark);
-                if (priorChangeTime !== undefined) {
-                    newestDeclarationFileContentChangedTime = newer(priorChangeTime, newestDeclarationFileContentChangedTime);
+                if (!isIncremental) {
+                    (outputTimeStampMap ||= getOutputTimeStampMap(state, projectPath)).set(path, now ||= getCurrentTime(state.host));

```

Comment:
All non incremental builds need to check with output files's time stamp so we are storing them in cache here.. 
For incremental build, only build info is sufficient so we dont need to modify or query the timestamps